miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/permission/MiniAppCustomPermissionCache.kt

```diff
@@ -17,24 +17,60 @@ internal class MiniAppCustomPermissionCache(context: Context) {
         "com.rakuten.tech.mobile.miniapp.custom.permissions.cache", Context.MODE_PRIVATE
     )
 
+    @VisibleForTesting
+    fun isDataExist(miniAppId: String) = prefs.contains(miniAppId)
+
     /**
      * Reads the grant results from SharedPreferences.
      * @param [miniAppId] the key provided to find the stored results per MiniApp
      * @return [MiniAppCustomPermission] an object to contain the results per MiniApp
      * if data has been stored in cache, otherwise default value.
      */
     fun readPermissions(miniAppId: String): MiniAppCustomPermission {
-        if (prefs.contains(miniAppId)) {
-            return try {
-                Gson().fromJson(
+        val defaultValue = defaultDeniedList(miniAppId)
+        return if (isDataExist(miniAppId)) {
+            try {
+                val cachedPermission: MiniAppCustomPermission = Gson().fromJson(
                     prefs.getString(miniAppId, ""),
                     object : TypeToken<MiniAppCustomPermission>() {}.type
                 )
+                val cachedPairs = cachedPermission.pairValues.toMutableList()
+
+                // detect any new change with comparing cached permissions and defaultDeniedList
+                // change means added new permission / removed existing permission from defaultDeniedList
```

Comment:
Thanks for adding this - I didn't think of this case.

In the future, I think this won't be a problem after we add a permissions manifest, so each mini app will need to add a manifest which has a list of the permissions they want to use. And then we can get the list of permissions just by reading the manifest.

Reply:
Yes, I am having a similar thought. We can gradually improve the feature by having the permissions manifest.